js/src/data/test/__helpers__/index.js

```diff
@@ -0,0 +1,105 @@
+/**
+ * External dependencies
+ */
+import { cloneDeep, set, get, isPlainObject } from 'lodash';
+
+/**
+ * Use native `Object.freeze()` to make an object immutable, recursively freeze each property which is of type object.
+ *
+ * Copied from [Object.freeze() of MDN]{@link https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/freeze}.
+ *
+ * @param {Object|Array} object The object or array to freeze.
+ * @return {Object|Array} The `object` that was passed-in this function.
+ */
+export function deepFreeze( object ) {
+	// Retrieve the property names defined on object
+	const propNames = Object.getOwnPropertyNames( object );
+
+	// Freeze properties before freezing self
+	for ( const name of propNames ) {
+		const value = object[ name ];
+		if ( value && typeof value === 'object' ) {
+			deepFreeze( value );
+		}
+	}
+
+	return Object.freeze( object );
+}
+
+function attachRef( object, ignore, checkingGroups = [], paths = [] ) {
+	const refKey = '__refForCheck';
+
+	for ( const [ name, value ] of Object.entries( object ) ) {
+		if ( isPlainObject( value ) ) {
+			const checkPaths = [ ...paths, name ];
+			attachRef( value, ignore, checkingGroups, checkPaths );
+
+			const currentPath = checkPaths.join( '.' );
+			if ( ! ignore.includes( currentPath ) ) {
+				const ref = {
+					whyFail: `If there's no mutation at \`state.${ currentPath }\`, it should be kept the same reference.`,
+				};
+				checkingGroups.push( {
+					ref,
+					refPath: `${ currentPath }.${ refKey }`,
+				} );
+				value[ refKey ] = ref;
+			}
+		}
+	}
+	return checkingGroups;
+}
```

Comment:
💅 
Missing docs

Reply:
Added by https://github.com/woocommerce/google-listings-and-ads/pull/1153/commits/9ca8505925788cd3b588e18df2efec8c65a8f748